File: src/main/java/com/mercor/assignment/scd/domain/core/model/SCDEntity.java
package com.mercor.assignment.scd.domain.core.model;

import java.util.Date;

/**
 * Base interface for all SCD entities
 */
public interface SCDEntity {

  String getId();

  void setId(String id);

  Integer getVersion();

  void setVersion(Integer version);

  String getUid();

  void setUid(String uid);

  Date getCreatedAt();

  void setCreatedAt(Date createdAt);

  Date getUpdatedAt();

  void setUpdatedAt(Date updatedAt);

  SCDEntity cloneForNewVersion(String uid, int version, Date now);
}

File: pkg/models/scd_entity.go
package models

import (
	"time"
)

// SCDEntity is the interface for all SCD entities
type SCDEntity interface {
	GetID() string
	SetID(string)
	
	GetVersion() int
	SetVersion(int)
	
	GetUID() string
	SetUID(string)
	
	GetCreatedAt() time.Time
	SetCreatedAt(time.Time)
	
	GetUpdatedAt() time.Time
	SetUpdatedAt(time.Time)
	
	TableName() string
	CloneForNewVersion(uid string, version int, now time.Time) SCDEntity
}
